eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/EventCriteriaBuilder.java

```diff
@@ -17,50 +17,93 @@
+ *          {@link EventCriteriaBuilderTagStage#withTags(String...)}.
+ *     </li>
+ * </ol>
+ * <p>
+ * The builder will return an {@link EventCriteria} instance that matches on the defined types and tags.
+ *
+ * @author Mitchell Herrijgers
+ * @see EventCriteria
+ * @since 5.0.0
+ */
+public final class EventCriteriaBuilder implements EventCriteriaBuilderEventTypeStage, EventCriteriaBuilderTagStage {
 
-    private final String[] types;
+    private final EventCriteria orCriteria;
+    private final Set<String> eventTypes = new HashSet<>();
 
-    static final EventCriteriaBuilder NO_TYPES = new EventCriteriaBuilder();
 
-    EventCriteriaBuilder(String... types) {
-        this.types = types;
+    private EventCriteriaBuilder(@Nullable EventCriteria orCriteria) {
+        this.orCriteria = orCriteria;
     }
 
-    @Override
-    public EventCriteria withTags(@Nonnull Set<Tag> tags) {
-        if (tags.isEmpty() && types.length == 0) {
-            return AnyEvent.INSTANCE;
-        }
-        return new DefaultEventCriteria(Set.of(types), tags);
+
+    /**
+     * Start building a new {@link EventCriteria} instance.
+     *
+     * @return The builder in the {@link EventCriteriaBuilderEventTypeStage event type stage}.
+     */
+    public static EventCriteriaBuilderEventTypeStage match() {
+        return new EventCriteriaBuilder(null);
     }
 
-    @Override
-    public EventCriteria withTags(@Nonnull Tag... tags) {
-        return withTags(Set.of(tags));
+    /**
+     * Start building a new {@link EventCriteria} instance that when constructed will be combined with the given
+     * {@code criteria} using an {@link EventCriteria#or(EventCriteria) OR operation}.
+     *
+     * @param criteria The criteria to combine with the new criteria using an OR operation.
+     * @return The builder in the {@link EventCriteriaBuilderEventTypeStage event type stage}.
+     */
+    public static EventCriteriaBuilderEventTypeStage or(EventCriteria criteria) {
+        return new EventCriteriaBuilder(criteria);
     }
 
-    @Override
-    public EventCriteria withTags(@Nonnull String... tagKeyValuePairs) {
-        if (tagKeyValuePairs.length % 2 != 0) {
-            throw new IllegalArgumentException("tagKeyValuePairs must have even length");
-        }
-        Set<Tag> tags = new HashSet<>();
-        for (int i = 0; i < tagKeyValuePairs.length; i = i + 2) {
-            tags.add(new Tag(tagKeyValuePairs[i], tagKeyValuePairs[i + 1]));
+    public EventCriteriaBuilderTagStage eventTypes(@Nonnull String... types) {
+        eventTypes.addAll(Arrays.asList(types));
+        return this;
+    }
+
+    public EventCriteriaBuilderTagStage anyEventType() {
+        return this;
+    }
+
+    public EventCriteria withTags(@Nonnull Set<Tag> tags) {
+        if (tags.isEmpty()) {
+            return withAnyTags();
         }
-        return withTags(tags);
+        FilteredEventCriteria criterion = new FilteredEventCriteria(eventTypes, tags);
```

Comment:
Super duper nit: my OCD triggers on the fact that the `FilteredEventCriteria` variable is called `criterion` her and is called `filteredEventCriteria` in the `withAnyTags` method :see_no_evil:

Reply:
Fair enough! Changing